From 3161fcfb187258a277ec5b236f01dccf83d57f31 Mon Sep 17 00:00:00 2001
From: Vladimir Barinov <vladimir.barinov@cogentembedded.com>
Date: Fri, 25 Sep 2020 18:54:32 +0300
Subject: [PATCH 02/11] uio stuff for BSP7

This is uio related changed for BSP7 (Yocto4.1.0)

Signed-off-by: Vladimir Barinov <vladimir.barinov@cogentembedded.com>
[valentine: removed generic stuff, added v3m es1 and v3h es2 uio changes]
Signed-off-by: Valentine Barshak <valentine.barshak@cogentembedded.com>
---
 arch/arm64/boot/dts/renesas/r8a77970-es1.dtsi |   31 +-
 arch/arm64/boot/dts/renesas/r8a77970.dtsi     |  505 +++++++++++-
 arch/arm64/boot/dts/renesas/r8a77980-es2.dtsi |    3 +-
 arch/arm64/boot/dts/renesas/r8a77980.dtsi     | 1014 +++++++++++++++++++++++--
 drivers/clk/renesas/r8a77970-cpg-mssr.c       |    4 +-
 drivers/clk/renesas/r8a77980-cpg-mssr.c       |    6 +-
 include/dt-bindings/clock/r8a77970-cpg-mssr.h |    2 +
 include/dt-bindings/clock/r8a77980-cpg-mssr.h |    3 +
 8 files changed, 1476 insertions(+), 92 deletions(-)

diff --git a/arch/arm64/boot/dts/renesas/r8a77970-es1.dtsi b/arch/arm64/boot/dts/renesas/r8a77970-es1.dtsi
index ca004a6..083770f 100644
--- a/arch/arm64/boot/dts/renesas/r8a77970-es1.dtsi
+++ b/arch/arm64/boot/dts/renesas/r8a77970-es1.dtsi
@@ -13,7 +13,7 @@
 / {
 	soc {
 		imp_distributer: imp_ctrl_00 {
-			compatible = "renesas,impx4-distributer";
+			compatible = "generic-uio";
 			reg = <0 0xffa00000 0 0x10000>;
 			interrupts = <GIC_SPI 281 IRQ_TYPE_LEVEL_HIGH>;
 			clocks = <&cpg CPG_MOD 830>;
@@ -23,91 +23,100 @@
 		};
 
 		imp_imp_00 {
-			compatible = "renesas,impx4-legacy";
+			compatible = "generic-uio";
 			reg = <0 0xff900000 0 0x20000>;
 			interrupt-parent = <&imp_distributer>;
 			interrupts = <0>;
 			clocks = <&cpg CPG_MOD 827>;
 			power-domains = <&sysc R8A77970_PD_A2IR0>;
+			ipmmu-id = <RCAR_IPMMU_IR>;
 		};
 
 		imp_imp_01 {
-			compatible = "renesas,impx4-legacy";
+			compatible = "generic-uio";
 			reg = <0 0xff920000 0 0x20000>;
 			interrupt-parent = <&imp_distributer>;
 			interrupts = <1>;
 			clocks = <&cpg CPG_MOD 826>;
 			power-domains = <&sysc R8A77970_PD_A2IR1>;
+			ipmmu-id = <RCAR_IPMMU_IR>;
 		};
 
 		imp_imp_02 {
-			compatible = "renesas,impx4-legacy";
+			compatible = "generic-uio";
 			reg = <0 0xff940000 0 0x20000>;
 			interrupt-parent = <&imp_distributer>;
 			interrupts = <2>;
 			clocks = <&cpg CPG_MOD 825>;
 			power-domains = <&sysc R8A77970_PD_A2DP>;
+			ipmmu-id = <RCAR_IPMMU_IR>;
 		};
 
 		imp_imp_03 {
-			compatible = "renesas,impx4-legacy";
+			compatible = "generic-uio";
 			reg = <0 0xff960000 0 0x20000>;
 			interrupt-parent = <&imp_distributer>;
 			interrupts = <3>;
 			clocks = <&cpg CPG_MOD 824>;
 			power-domains = <&sysc R8A77970_PD_A2CN>;
+			ipmmu-id = <RCAR_IPMMU_IR>;
 		};
 
 		imp_ocv_00 {
-			compatible = "renesas,impx4-shader";
+			compatible = "generic-uio";
 			reg = <0 0xff980000 0 0x10000>;
 			interrupt-parent = <&imp_distributer>;
 			interrupts = <4>;
 			clocks = <&cpg CPG_MOD 829>;
 			power-domains = <&sysc R8A77970_PD_A2SC0>;
+			ipmmu-id = <RCAR_IPMMU_IR>;
 		};
 
 		imp_ocv_01 {
-			compatible = "renesas,impx4-shader";
+			compatible = "generic-uio";
 			reg = <0 0xff990000 0 0x10000>;
 			interrupt-parent = <&imp_distributer>;
 			interrupts = <5>;
 			clocks = <&cpg CPG_MOD 828>;
 			power-domains = <&sysc R8A77970_PD_A2SC1>;
+			ipmmu-id = <RCAR_IPMMU_IR>;
 		};
 
 		imp_dma_00 {
-			compatible = "renesas,impx5-dmac";
+			compatible = "generic-uio";
 			reg = <0 0xffa10000 0 0x1000>;
 			interrupt-parent = <&imp_distributer>;
 			interrupts = <16>;
 			clocks = <&cpg CPG_MOD 830>;
 			power-domains = <&sysc R8A77970_PD_A3IR>;
+			ipmmu-id = <RCAR_IPMMU_IR>;
 		};
 
 		imp_dma_01 {
-			compatible = "renesas,impx5-dmac";
+			compatible = "generic-uio";
 			reg = <0 0xffa10000 0 0x1000>,
 			      <0 0xffa10800 0 0x0800>;
 			interrupt-parent = <&imp_distributer>;
 			interrupts = <17>;
 			clocks = <&cpg CPG_MOD 830>;
 			power-domains = <&sysc R8A77970_PD_A3IR>;
+			ipmmu-id = <RCAR_IPMMU_IR>;
 		};
 
 		imp_psc_00 {
-			compatible = "renesas,impx5+-psc";
+			compatible = "generic-uio";
 			reg = <0 0xffa20000 0 0x4000>;
 			interrupt-parent = <&imp_distributer>;
 			interrupts = <12>;
 			clocks = <&cpg CPG_MOD 830>;
 			power-domains = <&sysc R8A77970_PD_A3IR>;
+			ipmmu-id = <RCAR_IPMMU_IR>;
 		};
 
 		/delete-node/imp_cnn_00;
 
 		imp_mem_00 {
-			compatible = "renesas,impx4-memory";
+			compatible = "generic-uio";
 			reg = <0 0xed000000 0 0xe0000>;
 			clocks = <&cpg CPG_MOD 830>;
 			power-domains = <&sysc R8A77970_PD_A3IR>;
diff --git a/arch/arm64/boot/dts/renesas/r8a77970.dtsi b/arch/arm64/boot/dts/renesas/r8a77970.dtsi
index 32b410e..5f25dc7 100644
--- a/arch/arm64/boot/dts/renesas/r8a77970.dtsi
+++ b/arch/arm64/boot/dts/renesas/r8a77970.dtsi
@@ -10,6 +10,7 @@
 #include <dt-bindings/interrupt-controller/arm-gic.h>
 #include <dt-bindings/interrupt-controller/irq.h>
 #include <dt-bindings/power/r8a77970-sysc.h>
+#include <dt-bindings/ipmmu/rcar-ipmmu.h>
 
 / {
 	compatible = "renesas,r8a77970";
@@ -22,6 +23,9 @@
 		i2c2 = &i2c2;
 		i2c3 = &i2c3;
 		i2c4 = &i2c4;
+		isp0 = &isp0;
+		tisp0 = &tisp0;
+		cisp0 = &cisp0;
 	};
 
 	/* External CAN clock - to be overridden by boards that provide it */
@@ -843,15 +847,16 @@
 			status = "disabled";
 		};
 
-		vin0: video@e6ef0000 {
-			compatible = "renesas,vin-r8a77970";
+		vin0: vin_00@e6ef0000 {
+			compatible = "renesas,vin-r8a77970", "generic-uio";
 			reg = <0 0xe6ef0000 0 0x1000>;
 			interrupts = <GIC_SPI 188 IRQ_TYPE_LEVEL_HIGH>;
 			clocks = <&cpg CPG_MOD 811>;
 			power-domains = <&sysc R8A77970_PD_ALWAYS_ON>;
 			resets = <&cpg 811>;
 			renesas,id = <0>;
-			status = "disabled";
+			status = "enabled";
+			ipmmu-id = <RCAR_IPMMU_VI0>;
 
 			ports {
 				#address-cells = <1>;
@@ -871,15 +876,16 @@
 			};
 		};
 
-		vin1: video@e6ef1000 {
-			compatible = "renesas,vin-r8a77970";
+		vin1: vin_01@e6ef1000 {
+			compatible = "renesas,vin-r8a77970", "generic-uio";
 			reg = <0 0xe6ef1000 0 0x1000>;
 			interrupts = <GIC_SPI 189 IRQ_TYPE_LEVEL_HIGH>;
 			clocks = <&cpg CPG_MOD 810>;
 			power-domains = <&sysc R8A77970_PD_ALWAYS_ON>;
 			resets = <&cpg 810>;
 			renesas,id = <1>;
-			status = "disabled";
+			status = "enabled";
+			ipmmu-id = <RCAR_IPMMU_VI0>;
 
 			ports {
 				#address-cells = <1>;
@@ -899,15 +905,16 @@
 			};
 		};
 
-		vin2: video@e6ef2000 {
-			compatible = "renesas,vin-r8a77970";
+		vin2: vin_02@e6ef2000 {
+			compatible = "renesas,vin-r8a77970", "generic-uio";
 			reg = <0 0xe6ef2000 0 0x1000>;
 			interrupts = <GIC_SPI 190 IRQ_TYPE_LEVEL_HIGH>;
 			clocks = <&cpg CPG_MOD 809>;
 			power-domains = <&sysc R8A77970_PD_ALWAYS_ON>;
 			resets = <&cpg 809>;
 			renesas,id = <2>;
-			status = "disabled";
+			status = "enabled";
+			ipmmu-id = <RCAR_IPMMU_VI0>;
 
 			ports {
 				#address-cells = <1>;
@@ -927,15 +934,16 @@
 			};
 		};
 
-		vin3: video@e6ef3000 {
-			compatible = "renesas,vin-r8a77970";
+		vin3: vin_03@e6ef3000 {
+			compatible = "renesas,vin-r8a77970", "generic-uio";
 			reg = <0 0xe6ef3000 0 0x1000>;
 			interrupts = <GIC_SPI 191 IRQ_TYPE_LEVEL_HIGH>;
 			clocks = <&cpg CPG_MOD 808>;
 			power-domains = <&sysc R8A77970_PD_ALWAYS_ON>;
 			resets = <&cpg 808>;
 			renesas,id = <3>;
-			status = "disabled";
+			status = "enabled";
+			ipmmu-id = <RCAR_IPMMU_VI0>;
 
 			ports {
 				#address-cells = <1>;
@@ -955,6 +963,187 @@
 			};
 		};
 
+		imp_distributer: imp_ctrl_00 {
+			compatible = "generic-uio";
+			reg = <0 0xffa00000 0 0x10000>;
+			interrupts = <GIC_SPI 281 IRQ_TYPE_LEVEL_HIGH>;
+			clocks = <&cpg CPG_MOD 830>;
+			power-domains = <&sysc R8A77970_PD_A3IR>;
+			interrupt-controller;
+			#interrupt-cells = <1>;
+		};
+
+		imp_imp_00 {
+			compatible = "generic-uio";
+			reg = <0 0xff900000 0 0x20000>;
+			clocks = <&cpg CPG_MOD 827>;
+			power-domains = <&sysc R8A77970_PD_A2IR0>;
+			ipmmu-id = <RCAR_IPMMU_IR>;
+		};
+
+		imp_imp_01 {
+			compatible = "generic-uio";
+			reg = <0 0xff920000 0 0x20000>;
+			clocks = <&cpg CPG_MOD 827>;
+			power-domains = <&sysc R8A77970_PD_A2IR0>;
+			ipmmu-id = <RCAR_IPMMU_IR>;
+		};
+
+		imp_imp_02 {
+			compatible = "generic-uio";
+			reg = <0 0xff940000 0 0x20000>;
+			clocks = <&cpg CPG_MOD 826>;
+			power-domains = <&sysc R8A77970_PD_A2IR1>;
+			ipmmu-id = <RCAR_IPMMU_IR>;
+		};
+
+		imp_imp_03 {
+			compatible = "generic-uio";
+			reg = <0 0xff960000 0 0x20000>;
+			clocks = <&cpg CPG_MOD 826>;
+			power-domains = <&sysc R8A77970_PD_A2IR1>;
+			ipmmu-id = <RCAR_IPMMU_IR>;
+		};
+
+		imp_ocv_00 {
+			compatible = "generic-uio";
+			reg = <0 0xff980000 0 0x10000>;
+			clocks = <&cpg CPG_MOD 829>;
+			power-domains = <&sysc R8A77970_PD_A2SC0>;
+			ipmmu-id = <RCAR_IPMMU_IR>;
+		};
+
+		imp_ocv_01 {
+			compatible = "generic-uio";
+			reg = <0 0xff990000 0 0x10000>;
+			clocks = <&cpg CPG_MOD 828>;
+			power-domains = <&sysc R8A77970_PD_A2SC1>;
+			ipmmu-id = <RCAR_IPMMU_IR>;
+		};
+
+		imp_dma_00 {
+			compatible = "generic-uio";
+			reg = <0 0xffa10000 0 0x1000>;
+			clocks = <&cpg CPG_MOD 825>;
+			power-domains = <&sysc R8A77970_PD_A2DP>;
+			ipmmu-id = <RCAR_IPMMU_IR>;
+		};
+
+		imp_psc_00 {
+			compatible = "generic-uio";
+			reg = <0 0xffa20000 0 0x4000>;
+			clocks = <&cpg CPG_MOD 825>;
+			power-domains = <&sysc R8A77970_PD_A2DP>;
+			ipmmu-id = <RCAR_IPMMU_IR>;
+		};
+
+		imp_cnn_00 {
+			compatible = "generic-uio";
+			reg = <0 0xff9e0000 0 0x10000>;
+			clocks = <&cpg CPG_MOD 824>;
+			power-domains = <&sysc R8A77970_PD_A2CN>;
+			ipmmu-id = <RCAR_IPMMU_IR>;
+		};
+
+		imp_mem_00 {
+			compatible = "generic-uio";
+			reg = <0 0xed000000 0 0xe0000>;
+			clocks = <&cpg CPG_MOD 830>;
+			power-domains = <&sysc R8A77970_PD_A3IR>;
+		};
+
+		imr_v4l2_alloc: imr_alloc {
+			dma-coherent;
+		};
+
+		imrlx4_ch0: imr_00@fe860000 {
+			compatible = "generic-uio",
+				     "renesas,imr-lx4";
+			reg = <0 0xfe860000 0 0x2000>;
+			interrupts = <GIC_SPI 192 IRQ_TYPE_LEVEL_HIGH>;
+			clocks = <&cpg CPG_MOD 823>;
+			power-domains = <&sysc R8A77970_PD_ALWAYS_ON>;
+			resets = <&cpg 823>;
+			alloc-dev = <&imr_v4l2_alloc>;
+			ipmmu-id = <RCAR_IPMMU_VI0>;
+		};
+
+		imrlx4_ch1: imr_01@fe870000 {
+			compatible = "generic-uio",
+				     "renesas,imr-lx4";
+			reg = <0 0xfe870000 0 0x2000>;
+			interrupts = <GIC_SPI 193 IRQ_TYPE_LEVEL_HIGH>;
+			clocks = <&cpg CPG_MOD 822>;
+			power-domains = <&sysc R8A77970_PD_ALWAYS_ON>;
+			resets = <&cpg 822>;
+			alloc-dev = <&imr_v4l2_alloc>;
+			ipmmu-id = <RCAR_IPMMU_VI0>;
+		};
+
+		imrlx4_ch2: imr_02@fe880000 {
+			compatible = "generic-uio",
+				     "renesas,imr-lx4";
+			reg = <0 0xfe880000 0 0x2000>;
+			interrupts = <GIC_SPI 194 IRQ_TYPE_LEVEL_HIGH>;
+			clocks = <&cpg CPG_MOD 821>;
+			power-domains = <&sysc R8A77970_PD_ALWAYS_ON>;
+			resets = <&cpg 821>;
+			alloc-dev = <&imr_v4l2_alloc>;
+			ipmmu-id = <RCAR_IPMMU_VI0>;
+		};
+
+		imrlx4_ch3: imr_03@fe890000 {
+			compatible = "generic-uio",
+				     "renesas,imr-lx4";
+			reg = <0 0xfe890000 0 0x2000>;
+			interrupts = <GIC_SPI 195 IRQ_TYPE_LEVEL_HIGH>;
+			clocks = <&cpg CPG_MOD 820>;
+			power-domains = <&sysc R8A77970_PD_ALWAYS_ON>;
+			resets = <&cpg 820>;
+			alloc-dev = <&imr_v4l2_alloc>;
+			ipmmu-id = <RCAR_IPMMU_VI0>;
+		};
+
+		isp0: isp@fec00000 {
+			compatible = "renesas,isp-r8a77970";
+			reg = <0 0xfec00000 0 0x20000>,
+			      <0 0xfed00000 0 0x10000>;
+			interrupts = <GIC_SPI 26 IRQ_TYPE_LEVEL_HIGH>,
+				     <GIC_SPI 25 IRQ_TYPE_LEVEL_HIGH>;
+			clocks = <&cpg CPG_MOD 817>;
+			power-domains = <&sysc R8A77970_PD_ALWAYS_ON>;
+		};
+
+		tisp0: tisp_00 {
+			compatible = "generic-uio";
+			reg = <0 0xfed00000 0 0x10000>;
+			interrupts = <GIC_SPI 25 IRQ_TYPE_LEVEL_HIGH>;
+			clocks = <&cpg CPG_MOD 817>;
+			power-domains = <&sysc R8A77970_PD_ALWAYS_ON>;
+			ipmmu-id = <RCAR_IPMMU_VI0>;
+		};
+
+		cisp0: cisp_00 {
+			compatible = "generic-uio";
+			reg = <0 0xfec00000 0 0x20000>;
+			interrupts = <GIC_SPI 26 IRQ_TYPE_LEVEL_HIGH>;
+			clocks = <&cpg CPG_MOD 817>;
+			power-domains = <&sysc R8A77970_PD_ALWAYS_ON>;
+			ipmmu-id = <RCAR_IPMMU_VI0>;
+		};
+
+		sadc0: adc@ffce0000 {
+			compatible = "renesas,sadc-r8a77970";
+			reg = <0 0xffce0000 0 0x10000>;
+			interrupts = <GIC_SPI 295 IRQ_TYPE_LEVEL_HIGH>;
+			clocks = <&cpg CPG_CORE R8A77970_CLK_S2D4>,
+				 <&cpg CPG_CORE R8A77970_CLK_S2D2>,
+				 <&cpg CPG_MOD 503>;				/* RMSTPCR5/bit3:ADC */
+			clock-names = "phy_clk", "ip_clk", "adc";
+			power-domains = <&sysc R8A77970_PD_ALWAYS_ON>;
+			status = "disabled";
+		};
+
 		dmac1: dma-controller@e7300000 {
 			compatible = "renesas,dmac-r8a77970",
 				     "renesas,rcar-dmac";
@@ -1081,8 +1270,8 @@
 			resets = <&cpg 408>;
 		};
 
-		vspd0: vsp@fea20000 {
-			compatible = "renesas,vsp2";
+		vspd0: vsp_00@fea20000 {
+			compatible = "renesas,vsp2", "generic-uio";
 			reg = <0 0xfea20000 0 0x5000>;
 			interrupts = <GIC_SPI 169 IRQ_TYPE_LEVEL_HIGH>;
 			clocks = <&cpg CPG_MOD 623>;
@@ -1099,14 +1288,14 @@
 			resets = <&cpg 603>;
 		};
 
-		csi40: csi2@feaa0000 {
-			compatible = "renesas,r8a77970-csi2";
+		csi40: csi_00@feaa0000 {
+			compatible = "renesas,r8a77970-csi2", "generic-uio";
 			reg = <0 0xfeaa0000 0 0x10000>;
 			interrupts = <GIC_SPI 246 IRQ_TYPE_LEVEL_HIGH>;
 			clocks = <&cpg CPG_MOD 716>;
 			power-domains = <&sysc R8A77970_PD_ALWAYS_ON>;
 			resets = <&cpg 716>;
-			status = "disabled";
+			status = "enabled";
 
 			ports {
 				#address-cells = <1>;
@@ -1138,8 +1327,8 @@
 			};
 		};
 
-		du: display@feb00000 {
-			compatible = "renesas,du-r8a77970";
+		du: du_00@feb00000 {
+			compatible = "renesas,du-r8a77970", "generic-uio";
 			reg = <0 0xfeb00000 0 0x80000>;
 			interrupts = <GIC_SPI 256 IRQ_TYPE_LEVEL_HIGH>;
 			clocks = <&cpg CPG_MOD 724>;
@@ -1147,7 +1336,7 @@
 			power-domains = <&sysc R8A77970_PD_ALWAYS_ON>;
 			resets = <&cpg 724>;
 			vsps = <&vspd0>;
-			status = "disabled";
+			status = "enabled";
 
 			ports {
 				#address-cells = <1>;
@@ -1168,9 +1357,9 @@
 			};
 		};
 
-		lvds0: lvds-encoder@feb90000 {
-			compatible = "renesas,r8a77970-lvds";
-			reg = <0 0xfeb90000 0 0x14>;
+		lvds0: lvds_00@feb90000 {
+			compatible = "renesas,r8a77970-lvds", "generic-uio";
+			reg = <0 0xfeb90000 0 0x1000>;
 			clocks = <&cpg CPG_MOD 727>;
 			power-domains = <&sysc R8A77970_PD_ALWAYS_ON>;
 			resets = <&cpg 727>;
@@ -1199,6 +1388,276 @@
 			compatible = "renesas,prr";
 			reg = <0 0xfff00044 0 4>;
 		};
+
+		iccom_00 {
+			compatible = "renesas,iccom-rcar", "generic-uio";
+			interrupts = <0 224 4>;
+			reg = <0x0 0xe6260000 0x0 0x1000>;
+			iccom,reg-offset = <0x400 0x404 0x440 0x460>;
+			iccom,cta-memory = <0x0 0x47fc7000 0x0 0x2000>;
+			iccom,ack-timeout = <100>;
+			iccom,trg-timeout = <100>;
+		};
+
+		irq_iccom_00_01 {
+			compatible = "renesas,iccom-rcar", "generic-uio";
+			interrupts = <0 225 4>;
+			reg = <0x0 0xe6260000 0x0 0x1000>;
+			iccom,reg-offset = <0x408 0x40c 0x444 0x464>;
+			iccom,cta-memory = <0x0 0x47fc9000 0x0 0x2000>;
+			iccom,ack-timeout = <100>;
+			iccom,trg-timeout = <100>;
+		};
+
+		irq_iccom_00_02 {
+			compatible = "renesas,iccom-rcar", "generic-uio";
+			interrupts = <0 226 4>;
+			reg = <0x0 0xe6260000 0x0 0x1000>;
+			iccom,reg-offset = <0x410 0x414 0x448 0x468>;
+			iccom,cta-memory = <0x0 0x47fcb000 0x0 0x2000>;
+			iccom,ack-timeout = <100>;
+			iccom,trg-timeout = <100>;
+		};
+
+		irq_iccom_00_03 {
+			compatible = "renesas,iccom-rcar", "generic-uio";
+			interrupts = <0 227 4>;
+			reg = <0x0 0xe6260000 0x0 0x1000>;
+			iccom,reg-offset = <0x418 0x41c 0x44c 0x46c>;
+			iccom,cta-memory = <0x0 0x47fcd000 0x0 0x2000>;
+			iccom,ack-timeout = <100>;
+			iccom,trg-timeout = <100>;
+		};
+
+		irq_iccom_00_04 {
+			compatible = "renesas,iccom-rcar", "generic-uio";
+			interrupts = <0 228 4>;
+			reg = <0x0 0xe6260000 0x0 0x1000>;
+			iccom,reg-offset = <0x420 0x424 0x450 0x470>;
+			iccom,cta-memory = <0x0 0x47fcf000 0x0 0x2000>;
+			iccom,ack-timeout = <100>;
+			iccom,trg-timeout = <100>;
+		};
+
+		irq_iccom_00_05 {
+			compatible = "renesas,iccom-rcar", "generic-uio";
+			interrupts = <0 229 4>;
+			reg = <0x0 0xe6260000 0x0 0x1000>;
+			iccom,reg-offset = <0x428 0x42c 0x454 0x474>;
+			iccom,cta-memory = <0x0 0x47fd1000 0x0 0x2000>;
+			iccom,ack-timeout = <100>;
+			iccom,trg-timeout = <100>;
+		};
+
+		irq_iccom_00_06 {
+			compatible = "renesas,iccom-rcar", "generic-uio";
+			interrupts = <0 230 4>;
+			reg = <0x0 0xe6260000 0x0 0x1000>;
+			iccom,reg-offset = <0x430 0x434 0x458 0x478>;
+			iccom,cta-memory = <0x0 0x47fd3000 0x0 0x2000>;
+			iccom,ack-timeout = <100>;
+			iccom,trg-timeout = <100>;
+		};
+
+		irq_iccom_00_07 {
+			compatible = "renesas,iccom-rcar", "generic-uio";
+			interrupts = <0 231 4>;
+			reg = <0x0 0xe6260000 0x0 0x1000>;
+			iccom,reg-offset = <0x438 0x43c 0x45c 0x47c>;
+			iccom,cta-memory = <0x0 0x47fd5000 0x0 0x2000>;
+			iccom,ack-timeout = <100>;
+			iccom,trg-timeout = <100>;
+		};
+
+		irq_iccom_00_08 { /* iccom error 0 */
+			compatible = "generic-uio";
+			interrupts = <GIC_SPI 337 IRQ_TYPE_LEVEL_HIGH>;
+			reg = <0x0 0xe6260000 0x0 0x1000>;
+			power-domains = <&sysc R8A77970_PD_ALWAYS_ON>;
+		};
+
+		irq_iccom_00_09 { /* iccom error 1 */
+			compatible = "generic-uio";
+			interrupts = <GIC_SPI 338 IRQ_TYPE_LEVEL_HIGH>;
+			reg = <0x0 0xe6260000 0x0 0x1000>;
+			power-domains = <&sysc R8A77970_PD_ALWAYS_ON>;
+		};
+
+		irq_iccom_00_10 { /* iccom error 2 */
+			compatible = "generic-uio";
+			interrupts = <GIC_SPI 339 IRQ_TYPE_LEVEL_HIGH>;
+			reg = <0x0 0xe6260000 0x0 0x1000>;
+			power-domains = <&sysc R8A77970_PD_ALWAYS_ON>;
+		};
+
+		irq_iccom_00_11 { /* iccom error 3 */
+			compatible = "generic-uio";
+			interrupts = <GIC_SPI 340 IRQ_TYPE_LEVEL_HIGH>;
+			reg = <0x0 0xe6260000 0x0 0x1000>;
+			power-domains = <&sysc R8A77970_PD_ALWAYS_ON>;
+		};
+
+		irq_iccom_00_12 { /* iccom error 4 */
+			compatible = "generic-uio";
+			interrupts = <GIC_SPI 341 IRQ_TYPE_LEVEL_HIGH>;
+			reg = <0x0 0xe6260000 0x0 0x1000>;
+			power-domains = <&sysc R8A77970_PD_ALWAYS_ON>;
+		};
+
+		irq_iccom_00_13 { /* iccom error 5 */
+			compatible = "generic-uio";
+			interrupts = <GIC_SPI 342 IRQ_TYPE_LEVEL_HIGH>;
+			reg = <0x0 0xe6260000 0x0 0x1000>;
+			power-domains = <&sysc R8A77970_PD_ALWAYS_ON>;
+		};
+
+		irq_iccom_00_14 { /* iccom error 6 */
+			compatible = "generic-uio";
+			interrupts = <GIC_SPI 343 IRQ_TYPE_LEVEL_HIGH>;
+			reg = <0x0 0xe6260000 0x0 0x1000>;
+			power-domains = <&sysc R8A77970_PD_ALWAYS_ON>;
+		};
+
+		irq_iccom_00_15 { /* iccom error 7 */
+			compatible = "generic-uio";
+			interrupts = <GIC_SPI 344 IRQ_TYPE_LEVEL_HIGH>;
+			reg = <0x0 0xe6260000 0x0 0x1000>;
+			power-domains = <&sysc R8A77970_PD_ALWAYS_ON>;
+		};
+
+		irq_iccom_00_16 { /* iccom error 8 */
+			compatible = "generic-uio";
+			interrupts = <GIC_SPI 31 IRQ_TYPE_LEVEL_HIGH>;
+			reg = <0x0 0xe6260000 0x0 0x1000>;
+			power-domains = <&sysc R8A77970_PD_ALWAYS_ON>;
+		};
+
+		irq_iccom_00_17 { /* iccom error 9 */
+			compatible = "generic-uio";
+			interrupts = <GIC_SPI 347 IRQ_TYPE_LEVEL_HIGH>;
+			reg = <0x0 0xe6260000 0x0 0x1000>;
+			power-domains = <&sysc R8A77970_PD_ALWAYS_ON>;
+		};
+
+		irq_iccom_00_18 { /* iccom error 10 */
+			compatible = "generic-uio";
+			interrupts = <GIC_SPI 348 IRQ_TYPE_LEVEL_HIGH>;
+			reg = <0x0 0xe6260000 0x0 0x1000>;
+			power-domains = <&sysc R8A77970_PD_ALWAYS_ON>;
+		};
+
+		irq_iccom_00_19 { /* iccom error 11 */
+			compatible = "generic-uio";
+			interrupts = <GIC_SPI 349 IRQ_TYPE_LEVEL_HIGH>;
+			reg = <0x0 0xe6260000 0x0 0x1000>;
+			power-domains = <&sysc R8A77970_PD_ALWAYS_ON>;
+		};
+
+		qos@e67e0000 {
+			compatible = "renesas,qos";
+			reg = <0 0xe67e0000 0 0x10090>;
+		};
+
+		ivcp1e_00 {
+			compatible = "generic-uio";
+			reg = <0 0xfea00000 0 0x10000>;
+			clocks = <&cpg CPG_MOD 127>;
+			power-domains = <&sysc R8A77970_PD_ALWAYS_ON>;
+			interrupts = <GIC_SPI 300 IRQ_TYPE_LEVEL_HIGH>;
+			ipmmu-id = <RCAR_IPMMU_VI0>;
+		};
+
+		irq_ivcp1e_00_01 {
+			compatible = "generic-uio";
+			reg = <0 0xfea00000 0 0x10000>;
+			clocks = <&cpg CPG_MOD 127>;
+			power-domains = <&sysc R8A77970_PD_ALWAYS_ON>;
+			interrupts = <GIC_SPI 170 IRQ_TYPE_LEVEL_HIGH>;
+		};
+
+		rtt_fbc_post {
+			compatible = "generic-uio";
+			reg = <0 0xff830000 0 0x10000>;
+			clocks = <&cpg CPG_CORE R8A77970_CLK_POST>;
+			power-domains = <&sysc R8A77970_PD_ALWAYS_ON>;
+		};
+
+		rtt_fbc_post2 {
+			compatible = "generic-uio";
+			reg = <0 0xff830000 0 0x10000>;
+			clocks = <&cpg CPG_CORE R8A77970_CLK_POST2>;
+			power-domains = <&sysc R8A77970_PD_ALWAYS_ON>;
+		};
+
+		rtt_fbc_finish {
+			compatible = "generic-uio";
+			reg = <0 0xff830000 0 0x10000>;
+			power-domains = <&sysc R8A77970_PD_ALWAYS_ON>;
+			interrupts = <GIC_SPI 14 IRQ_TYPE_LEVEL_HIGH>;
+		};
+
+		rtt_fbc_pre {
+			compatible = "generic-uio";
+			reg = <0 0xff830000 0 0x10000>;
+			power-domains = <&sysc R8A77970_PD_ALWAYS_ON>;
+			interrupts = <GIC_SPI 15 IRQ_TYPE_LEVEL_HIGH>;
+		};
+
+		imr_rtt_ims_00 {
+			compatible = "generic-uio";
+			reg = <0 0xfff80000 0 0x1000>;
+			clocks = <&cpg CPG_MOD 823>;
+			power-domains = <&sysc R8A77970_PD_ALWAYS_ON>;
+		};
+
+		imr_rtt_ims_01 {
+			compatible = "generic-uio";
+			reg = <0 0xfff90000 0 0x1000>;
+			clocks = <&cpg CPG_MOD 821>;
+			power-domains = <&sysc R8A77970_PD_ALWAYS_ON>;
+		};
+
+		imp_rtt_imp_00 {
+			compatible = "generic-uio";
+			reg = <0 0xff8d0000 0 0x1000>;
+			clocks = <&cpg CPG_MOD 827>;
+			power-domains = <&sysc R8A77970_PD_A2IR0>;
+		};
+
+		imp_rtt_imp_01 {
+			compatible = "generic-uio";
+			reg = <0 0xfff50000 0 0x1000>;
+			clocks = <&cpg CPG_MOD 826>;
+			power-domains = <&sysc R8A77970_PD_A2IR1>;
+		};
+
+		imp_rtt_ocv_00 {
+			compatible = "generic-uio";
+			reg = <0 0xfffa0000 0 0x1000>;
+			clocks = <&cpg CPG_MOD 829>;
+			power-domains = <&sysc R8A77970_PD_A2SC0>;
+		};
+
+		imp_rtt_ocv_01 {
+			compatible = "generic-uio";
+			reg = <0 0xfffb0000 0 0x1000>;
+			clocks = <&cpg CPG_MOD 828>;
+			power-domains = <&sysc R8A77970_PD_A2SC1>;
+		};
+
+		imp_rtt_dp_00 {
+			compatible = "generic-uio";
+			reg = <0 0xfff60000 0 0x1000>;
+			clocks = <&cpg CPG_MOD 825>;
+			power-domains = <&sysc R8A77970_PD_A2DP>;
+		};
+
+		imp_rtt_cnn_00 {
+			compatible = "generic-uio";
+			reg = <0 0xfff70000 0 0x1000>;
+			clocks = <&cpg CPG_MOD 824>;
+			power-domains = <&sysc R8A77970_PD_A2CN>;
+		};
 	};
 
 	thermal-zones {
diff --git a/arch/arm64/boot/dts/renesas/r8a77980-es2.dtsi b/arch/arm64/boot/dts/renesas/r8a77980-es2.dtsi
index 615ec3b..0616584 100644
--- a/arch/arm64/boot/dts/renesas/r8a77980-es2.dtsi
+++ b/arch/arm64/boot/dts/renesas/r8a77980-es2.dtsi
@@ -18,10 +18,11 @@
 		/delete-node/imp_psc_01;
 
 		imp_slimdma_00 {
-			compatible = "renesas,slim-impx5-dmac";
+			compatible = "generic-uio";
 			reg = <0 0xffa18000 0 0x1000>;
 			clocks = <&cpg CPG_MOD 524>;
 			power-domains = <&sysc R8A77980_PD_A2DP1>;
+			ipmmu-id = <RCAR_IPMMU_IR>;
 		};
 
 		/* TBD: can fd channel 3 */
diff --git a/arch/arm64/boot/dts/renesas/r8a77980.dtsi b/arch/arm64/boot/dts/renesas/r8a77980.dtsi
index 6313ffe..de049b3 100644
--- a/arch/arm64/boot/dts/renesas/r8a77980.dtsi
+++ b/arch/arm64/boot/dts/renesas/r8a77980.dtsi
@@ -10,6 +10,7 @@
 #include <dt-bindings/interrupt-controller/irq.h>
 #include <dt-bindings/interrupt-controller/arm-gic.h>
 #include <dt-bindings/power/r8a77980-sysc.h>
+#include <dt-bindings/ipmmu/rcar-ipmmu.h>
 
 / {
 	compatible = "renesas,r8a77980";
@@ -23,6 +24,12 @@
 		i2c3 = &i2c3;
 		i2c4 = &i2c4;
 		i2c5 = &i2c5;
+		isp0 = &isp0;
+		isp1 = &isp1;
+		tisp0 = &tisp0;
+		tisp1 = &tisp1;
+		cisp0 = &cisp0;
+		cisp1 = &cisp1;
 	};
 
 	/* External CAN clock - to be overridden by boards that provide it */
@@ -886,15 +893,16 @@
 			status = "disabled";
 		};
 
-		vin0: video@e6ef0000 {
-			compatible = "renesas,vin-r8a77980";
+		vin0: vin_00@e6ef0000 {
+			compatible = "renesas,vin-r8a77980", "generic-uio";
 			reg = <0 0xe6ef0000 0 0x1000>;
 			interrupts = <GIC_SPI 188 IRQ_TYPE_LEVEL_HIGH>;
 			clocks = <&cpg CPG_MOD 811>;
 			power-domains = <&sysc R8A77980_PD_ALWAYS_ON>;
 			resets = <&cpg 811>;
 			renesas,id = <0>;
-			status = "disabled";
+			status = "enabled";
+			ipmmu-id = <RCAR_IPMMU_VI0>;
 
 			ports {
 				#address-cells = <1>;
@@ -914,15 +922,16 @@
 			};
 		};
 
-		vin1: video@e6ef1000 {
-			compatible = "renesas,vin-r8a77980";
+		vin1: vin_01@e6ef1000 {
+			compatible = "renesas,vin-r8a77980", "generic-uio";
 			reg = <0 0xe6ef1000 0 0x1000>;
 			interrupts = <GIC_SPI 189 IRQ_TYPE_LEVEL_HIGH>;
 			clocks = <&cpg CPG_MOD 810>;
 			power-domains = <&sysc R8A77980_PD_ALWAYS_ON>;
-			status = "disabled";
+			status = "enabled";
 			renesas,id = <1>;
 			resets = <&cpg 810>;
+			ipmmu-id = <RCAR_IPMMU_VI0>;
 
 			ports {
 				#address-cells = <1>;
@@ -942,15 +951,16 @@
 			};
 		};
 
-		vin2: video@e6ef2000 {
-			compatible = "renesas,vin-r8a77980";
+		vin2: vin_02@e6ef2000 {
+			compatible = "renesas,vin-r8a77980", "generic-uio";
 			reg = <0 0xe6ef2000 0 0x1000>;
 			interrupts = <GIC_SPI 190 IRQ_TYPE_LEVEL_HIGH>;
 			clocks = <&cpg CPG_MOD 809>;
 			power-domains = <&sysc R8A77980_PD_ALWAYS_ON>;
 			resets = <&cpg 809>;
 			renesas,id = <2>;
-			status = "disabled";
+			status = "enabled";
+			ipmmu-id = <RCAR_IPMMU_VI0>;
 
 			ports {
 				#address-cells = <1>;
@@ -970,15 +980,16 @@
 			};
 		};
 
-		vin3: video@e6ef3000 {
-			compatible = "renesas,vin-r8a77980";
+		vin3: vin_03@e6ef3000 {
+			compatible = "renesas,vin-r8a77980", "generic-uio";
 			reg = <0 0xe6ef3000 0 0x1000>;
 			interrupts = <GIC_SPI 191 IRQ_TYPE_LEVEL_HIGH>;
 			clocks = <&cpg CPG_MOD 808>;
 			power-domains = <&sysc R8A77980_PD_ALWAYS_ON>;
 			resets = <&cpg 808>;
 			renesas,id = <3>;
-			status = "disabled";
+			status = "enabled";
+			ipmmu-id = <RCAR_IPMMU_VI0>;
 
 			ports {
 				#address-cells = <1>;
@@ -998,15 +1009,16 @@
 			};
 		};
 
-		vin4: video@e6ef4000 {
-			compatible = "renesas,vin-r8a77980";
+		vin4: vin_04@e6ef4000 {
+			compatible = "renesas,vin-r8a77980", "generic-uio";
 			reg = <0 0xe6ef4000 0 0x1000>;
 			interrupts = <GIC_SPI 174 IRQ_TYPE_LEVEL_HIGH>;
 			clocks = <&cpg CPG_MOD 807>;
 			power-domains = <&sysc R8A77980_PD_ALWAYS_ON>;
 			resets = <&cpg 807>;
 			renesas,id = <4>;
-			status = "disabled";
+			status = "enabled";
+			ipmmu-id = <RCAR_IPMMU_VI0>;
 
 			ports {
 				#address-cells = <1>;
@@ -1026,15 +1038,16 @@
 			};
 		};
 
-		vin5: video@e6ef5000 {
-			compatible = "renesas,vin-r8a77980";
+		vin5: vin_05@e6ef5000 {
+			compatible = "renesas,vin-r8a77980", "generic-uio";
 			reg = <0 0xe6ef5000 0 0x1000>;
 			interrupts = <GIC_SPI 175 IRQ_TYPE_LEVEL_HIGH>;
 			clocks = <&cpg CPG_MOD 806>;
 			power-domains = <&sysc R8A77980_PD_ALWAYS_ON>;
 			resets = <&cpg 806>;
 			renesas,id = <5>;
-			status = "disabled";
+			status = "enabled";
+			ipmmu-id = <RCAR_IPMMU_VI0>;
 
 			ports {
 				#address-cells = <1>;
@@ -1054,15 +1067,16 @@
 			};
 		};
 
-		vin6: video@e6ef6000 {
-			compatible = "renesas,vin-r8a77980";
+		vin6: vin_06@e6ef6000 {
+			compatible = "renesas,vin-r8a77980", "generic-uio";
 			reg = <0 0xe6ef6000 0 0x1000>;
 			interrupts = <GIC_SPI 176 IRQ_TYPE_LEVEL_HIGH>;
 			clocks = <&cpg CPG_MOD 805>;
 			power-domains = <&sysc R8A77980_PD_ALWAYS_ON>;
 			resets = <&cpg 805>;
 			renesas,id = <6>;
-			status = "disabled";
+			status = "enabled";
+			ipmmu-id = <RCAR_IPMMU_VI0>;
 
 			ports {
 				#address-cells = <1>;
@@ -1082,15 +1096,16 @@
 			};
 		};
 
-		vin7: video@e6ef7000 {
-			compatible = "renesas,vin-r8a77980";
+		vin7: vin_07@e6ef7000 {
+			compatible = "renesas,vin-r8a77980", "generic-uio";
 			reg = <0 0xe6ef7000 0 0x1000>;
 			interrupts = <GIC_SPI 171 IRQ_TYPE_LEVEL_HIGH>;
 			clocks = <&cpg CPG_MOD 804>;
 			power-domains = <&sysc R8A77980_PD_ALWAYS_ON>;
 			resets = <&cpg 804>;
 			renesas,id = <7>;
-			status = "disabled";
+			status = "enabled";
+			ipmmu-id = <RCAR_IPMMU_VI0>;
 
 			ports {
 				#address-cells = <1>;
@@ -1110,8 +1125,8 @@
 			};
 		};
 
-		vin8: video@e6ef8000 {
-			compatible = "renesas,vin-r8a77980";
+		vin8: vin_08@e6ef8000 {
+			compatible = "renesas,vin-r8a77980", "generic-uio";
 			reg = <0 0xe6ef8000 0 0x1000>;
 			interrupts = <GIC_SPI 268 IRQ_TYPE_LEVEL_HIGH>;
 			clocks = <&cpg CPG_MOD 628>;
@@ -1119,10 +1134,11 @@
 			resets = <&cpg 628>;
 			renesas,id = <8>;
 			status = "disabled";
+			ipmmu-id = <RCAR_IPMMU_VI0>;
 		};
 
-		vin9: video@e6ef9000 {
-			compatible = "renesas,vin-r8a77980";
+		vin9: vin_09@e6ef9000 {
+			compatible = "renesas,vin-r8a77980", "generic-uio";
 			reg = <0 0xe6ef9000 0 0x1000>;
 			interrupts = <GIC_SPI 269 IRQ_TYPE_LEVEL_HIGH>;
 			clocks = <&cpg CPG_MOD 627>;
@@ -1130,10 +1146,11 @@
 			resets = <&cpg 627>;
 			renesas,id = <9>;
 			status = "disabled";
+			ipmmu-id = <RCAR_IPMMU_VI0>;
 		};
 
-		vin10: video@e6efa000 {
-			compatible = "renesas,vin-r8a77980";
+		vin10: vin_10@e6efa000 {
+			compatible = "renesas,vin-r8a77980", "generic-uio";
 			reg = <0 0xe6efa000 0 0x1000>;
 			interrupts = <GIC_SPI 289 IRQ_TYPE_LEVEL_HIGH>;
 			clocks = <&cpg CPG_MOD 625>;
@@ -1141,10 +1158,11 @@
 			resets = <&cpg 625>;
 			renesas,id = <10>;
 			status = "disabled";
+			ipmmu-id = <RCAR_IPMMU_VI0>;
 		};
 
-		vin11: video@e6efb000 {
-			compatible = "renesas,vin-r8a77980";
+		vin11: vin_11@e6efb000 {
+			compatible = "renesas,vin-r8a77980", "generic-uio";
 			reg = <0 0xe6efb000 0 0x1000>;
 			interrupts = <GIC_SPI 296 IRQ_TYPE_LEVEL_HIGH>;
 			clocks = <&cpg CPG_MOD 618>;
@@ -1152,10 +1170,11 @@
 			resets = <&cpg 618>;
 			renesas,id = <11>;
 			status = "disabled";
+			ipmmu-id = <RCAR_IPMMU_VI0>;
 		};
 
-		vin12: video@e6efc000 {
-			compatible = "renesas,vin-r8a77980";
+		vin12: vin_12@e6efc000 {
+			compatible = "renesas,vin-r8a77980", "generic-uio";
 			reg = <0 0xe6efc000 0 0x1000>;
 			interrupts = <GIC_SPI 298 IRQ_TYPE_LEVEL_HIGH>;
 			clocks = <&cpg CPG_MOD 612>;
@@ -1163,10 +1182,11 @@
 			resets = <&cpg 612>;
 			renesas,id = <12>;
 			status = "disabled";
+			ipmmu-id = <RCAR_IPMMU_VI0>;
 		};
 
-		vin13: video@e6efd000 {
-			compatible = "renesas,vin-r8a77980";
+		vin13: vin_13@e6efd000 {
+			compatible = "renesas,vin-r8a77980", "generic-uio";
 			reg = <0 0xe6efd000 0 0x1000>;
 			interrupts = <GIC_SPI 299 IRQ_TYPE_LEVEL_HIGH>;
 			clocks = <&cpg CPG_MOD 608>;
@@ -1174,10 +1194,11 @@
 			resets = <&cpg 608>;
 			renesas,id = <13>;
 			status = "disabled";
+			ipmmu-id = <RCAR_IPMMU_VI0>;
 		};
 
-		vin14: video@e6efe000 {
-			compatible = "renesas,vin-r8a77980";
+		vin14: vin_14@e6efe000 {
+			compatible = "renesas,vin-r8a77980", "generic-uio";
 			reg = <0 0xe6efe000 0 0x1000>;
 			interrupts = <GIC_SPI 301 IRQ_TYPE_LEVEL_HIGH>;
 			clocks = <&cpg CPG_MOD 605>;
@@ -1185,10 +1206,11 @@
 			resets = <&cpg 605>;
 			renesas,id = <14>;
 			status = "disabled";
+			ipmmu-id = <RCAR_IPMMU_VI0>;
 		};
 
-		vin15: video@e6eff000 {
-			compatible = "renesas,vin-r8a77980";
+		vin15: vin_15@e6eff000 {
+			compatible = "renesas,vin-r8a77980", "generic-uio";
 			reg = <0 0xe6eff000 0 0x1000>;
 			interrupts = <GIC_SPI 302 IRQ_TYPE_LEVEL_HIGH>;
 			clocks = <&cpg CPG_MOD 604>;
@@ -1196,6 +1218,485 @@
 			resets = <&cpg 604>;
 			renesas,id = <15>;
 			status = "disabled";
+			ipmmu-id = <RCAR_IPMMU_VI0>;
+		};
+
+		imp_distributer: imp_ctrl_00 {
+			compatible = "generic-uio";
+			reg = <0 0xffa00000 0 0x10000>;
+			interrupts = <GIC_SPI 281 IRQ_TYPE_LEVEL_HIGH>;
+			clocks = <&cpg CPG_MOD 830>;
+			power-domains = <&sysc R8A77980_PD_A3IR>;
+			interrupt-controller;
+			#interrupt-cells = <1>;
+		};
+
+		irq_imp_ctrl_00_01 {
+			compatible = "generic-uio";
+			reg = <0 0xffa00000 0 0x10000>;
+			interrupts = <GIC_SPI 282 IRQ_TYPE_LEVEL_HIGH>;
+			clocks = <&cpg CPG_MOD 830>;
+			power-domains = <&sysc R8A77980_PD_A3IR>;
+		};
+
+		irq_imp_ctrl_00_02 {
+			compatible = "generic-uio";
+			reg = <0 0xffa00000 0 0x10000>;
+			interrupts = <GIC_SPI 283 IRQ_TYPE_LEVEL_HIGH>;
+			clocks = <&cpg CPG_MOD 830>;
+			power-domains = <&sysc R8A77980_PD_A3IR>;
+		};
+
+		irq_imp_ctrl_00_03 {
+			compatible = "generic-uio";
+			reg = <0 0xffa00000 0 0x10000>;
+			interrupts = <GIC_SPI 284 IRQ_TYPE_LEVEL_HIGH>;
+			clocks = <&cpg CPG_MOD 830>;
+			power-domains = <&sysc R8A77980_PD_A3IR>;
+		};
+
+		irq_imp_ctrl_00_04 {
+			compatible = "generic-uio";
+			reg = <0 0xffa00000 0 0x10000>;
+			interrupts = <GIC_SPI 285 IRQ_TYPE_LEVEL_HIGH>;
+			clocks = <&cpg CPG_MOD 830>;
+			power-domains = <&sysc R8A77980_PD_A3IR>;
+		};
+
+		imp_imp_00 {
+			compatible = "generic-uio";
+			reg = <0 0xff900000 0 0x20000>;
+			clocks = <&cpg CPG_MOD 827>;
+			power-domains = <&sysc R8A77980_PD_A2IR0>;
+			ipmmu-id = <RCAR_IPMMU_IR>;
+		};
+
+		imp_imp_01 {
+			compatible = "generic-uio";
+			reg = <0 0xff920000 0 0x20000>;
+			clocks = <&cpg CPG_MOD 826>;
+			power-domains = <&sysc R8A77980_PD_A2IR1>;
+			ipmmu-id = <RCAR_IPMMU_IR>;
+		};
+
+		imp_imp_02 {
+			compatible = "generic-uio";
+			reg = <0 0xff940000 0 0x20000>;
+			clocks = <&cpg CPG_MOD 825>;
+			power-domains = <&sysc R8A77980_PD_A2IR2>;
+			ipmmu-id = <RCAR_IPMMU_IR>;
+		};
+
+		imp_imp_03 {
+			compatible = "generic-uio";
+			reg = <0 0xff960000 0 0x20000>;
+			clocks = <&cpg CPG_MOD 824>;
+			power-domains = <&sysc R8A77980_PD_A2IR3>;
+			ipmmu-id = <RCAR_IPMMU_IR>;
+		};
+
+		imp_imp_04 {
+			compatible = "generic-uio";
+			reg = <0 0xffa80000 0 0x20000>;
+			clocks = <&cpg CPG_MOD 521>;
+			power-domains = <&sysc R8A77980_PD_A2IR4>;
+			ipmmu-id = <RCAR_IPMMU_IR>;
+		};
+
+		imp_slim_00 {
+			compatible = "generic-uio";
+			reg = <0 0xff9c0000 0 0x10000>;
+			clocks = <&cpg CPG_MOD 500>;
+			power-domains = <&sysc R8A77980_PD_A2IR5>;
+			ipmmu-id = <RCAR_IPMMU_IR>;
+		};
+
+		imp_ocv_00 {
+			compatible = "generic-uio";
+			reg = <0 0xff980000 0 0x10000>;
+			clocks = <&cpg CPG_MOD 829>;
+			power-domains = <&sysc R8A77980_PD_A2SC0>;
+			ipmmu-id = <RCAR_IPMMU_IR>;
+		};
+
+		imp_ocv_01 {
+			compatible = "generic-uio";
+			reg = <0 0xff990000 0 0x10000>;
+			clocks = <&cpg CPG_MOD 828>;
+			power-domains = <&sysc R8A77980_PD_A2SC1>;
+			ipmmu-id = <RCAR_IPMMU_IR>;
+		};
+
+		imp_ocv_02 {
+			compatible = "generic-uio";
+			reg = <0 0xff9a0000 0 0x10000>;
+			clocks = <&cpg CPG_MOD 531>;
+			power-domains = <&sysc R8A77980_PD_A2SC2>;
+			ipmmu-id = <RCAR_IPMMU_IR>;
+		};
+
+		imp_ocv_03 {
+			compatible = "generic-uio";
+			reg = <0 0xff9b0000 0 0x10000>;
+			clocks = <&cpg CPG_MOD 529>;
+			power-domains = <&sysc R8A77980_PD_A2SC3>;
+			ipmmu-id = <RCAR_IPMMU_IR>;
+		};
+
+		imp_ocv_04 {
+			compatible = "generic-uio";
+			reg = <0 0xffa40000 0 0x10000>;
+			clocks = <&cpg CPG_MOD 528>;
+			power-domains = <&sysc R8A77980_PD_A2SC4>;
+			ipmmu-id = <RCAR_IPMMU_IR>;
+		};
+
+		imp_dma_00 {
+			compatible = "generic-uio";
+			reg = <0 0xffa10000 0 0x1000>;
+			clocks = <&cpg CPG_MOD 527>;
+			power-domains = <&sysc R8A77980_PD_A2DP0>;
+			ipmmu-id = <RCAR_IPMMU_IR>;
+		};
+
+		imp_dma_01 {
+			compatible = "generic-uio";
+			reg = <0 0xffa14000 0 0x1000>;
+			clocks = <&cpg CPG_MOD 526>;
+			power-domains = <&sysc R8A77980_PD_A2DP1>;
+			ipmmu-id = <RCAR_IPMMU_IR>;
+		};
+
+		imp_psc_00 {
+			compatible = "generic-uio";
+			reg = <0 0xffa20000 0 0x4000>;
+			clocks = <&cpg CPG_MOD 525>;
+			power-domains = <&sysc R8A77980_PD_A2DP0>;
+			ipmmu-id = <RCAR_IPMMU_IR>;
+		};
+
+		imp_psc_01 {
+			compatible = "generic-uio";
+			reg = <0 0xffa24000 0 0x4000>;
+			clocks = <&cpg CPG_MOD 524>;
+			power-domains = <&sysc R8A77980_PD_A2DP1>;
+			ipmmu-id = <RCAR_IPMMU_IR>;
+		};
+
+		imp_cnn_00 {
+			compatible = "generic-uio";
+			reg = <0 0xff9e0000 0 0x10000>;
+			clocks = <&cpg CPG_MOD 831>;
+			power-domains = <&sysc R8A77980_PD_A2CN>;
+			ipmmu-id = <RCAR_IPMMU_IR>;
+		};
+
+		imp_mem_00 {
+			compatible = "generic-uio";
+			reg = <0 0xed000000 0 0x200000>;
+			clocks = <&cpg CPG_MOD 830>;
+			power-domains = <&sysc R8A77980_PD_A3IR>;
+		};
+
+		imr_v4l2_alloc: imr_alloc {
+			dma-coherent;
+		};
+
+		imrlx4_ch0: imr_00@fe860000 {
+			compatible = "generic-uio",
+				     "renesas,imr-lx4";
+			reg = <0 0xfe860000 0 0x2000>;
+			interrupts = <GIC_SPI 192 IRQ_TYPE_LEVEL_HIGH>;
+			clocks = <&cpg CPG_MOD 823>;
+			power-domains = <&sysc R8A77980_PD_ALWAYS_ON>;
+			resets = <&cpg 823>;
+			alloc-dev = <&imr_v4l2_alloc>;
+			ipmmu-id = <RCAR_IPMMU_VC0>;
+		};
+
+		imrlx4_ch1: imr_01@fe870000 {
+			compatible = "generic-uio",
+				     "renesas,imr-lx4";
+			reg = <0 0xfe870000 0 0x2000>;
+			interrupts = <GIC_SPI 193 IRQ_TYPE_LEVEL_HIGH>;
+			clocks = <&cpg CPG_MOD 822>;
+			power-domains = <&sysc R8A77980_PD_ALWAYS_ON>;
+			resets = <&cpg 822>;
+			alloc-dev = <&imr_v4l2_alloc>;
+			ipmmu-id = <RCAR_IPMMU_VC0>;
+		};
+
+		imrlx4_ch2: imr_02@fe880000 {
+			compatible = "generic-uio",
+				     "renesas,imr-lx4";
+			reg = <0 0xfe880000 0 0x2000>;
+			interrupts = <GIC_SPI 194 IRQ_TYPE_LEVEL_HIGH>;
+			clocks = <&cpg CPG_MOD 821>;
+			power-domains = <&sysc R8A77980_PD_ALWAYS_ON>;
+			resets = <&cpg 821>;
+			alloc-dev = <&imr_v4l2_alloc>;
+			ipmmu-id = <RCAR_IPMMU_VC0>;
+		};
+
+		imrlx4_ch3: imr_03@fe890000 {
+			compatible = "generic-uio",
+				     "renesas,imr-lx4";
+			reg = <0 0xfe890000 0 0x2000>;
+			interrupts = <GIC_SPI 195 IRQ_TYPE_LEVEL_HIGH>;
+			clocks = <&cpg CPG_MOD 820>;
+			power-domains = <&sysc R8A77980_PD_ALWAYS_ON>;
+			resets = <&cpg 820>;
+			alloc-dev = <&imr_v4l2_alloc>;
+			ipmmu-id = <RCAR_IPMMU_VC0>;
+		};
+
+		imrlx4_ch4: imr_04@fe8a0000 {
+			compatible = "generic-uio",
+				     "renesas,imr-lx4";
+			reg = <0 0xfe8a0000 0 0x2000>;
+			interrupts = <GIC_SPI 254 IRQ_TYPE_LEVEL_HIGH>;
+			clocks = <&cpg CPG_MOD 707>;
+			power-domains = <&sysc R8A77980_PD_ALWAYS_ON>;
+			rse;
+			alloc-dev = <&imr_v4l2_alloc>;
+			ipmmu-id = <RCAR_IPMMU_VC0>;
+		};
+
+		imrlx4_ch5: imr_05@fe8b0000 {
+			compatible = "generic-uio",
+				     "renesas,imr-lx4";
+			reg = <0 0xfe8b0000 0 0x2000>;
+			interrupts = <GIC_SPI 255 IRQ_TYPE_LEVEL_HIGH>;
+			clocks = <&cpg CPG_MOD 706>;
+			power-domains = <&sysc R8A77980_PD_ALWAYS_ON>;
+			rse;
+			alloc-dev = <&imr_v4l2_alloc>;
+			ipmmu-id = <RCAR_IPMMU_VC0>;
+		};
+
+		stv_00 {
+			compatible = "generic-uio";
+			reg = <0 0xe7a00000 0 0x10000>;
+			interrupts = <GIC_SPI 238 IRQ_TYPE_LEVEL_HIGH>;
+			clocks = <&cpg CPG_MOD 101>;
+			power-domains = <&sysc R8A77980_PD_A3VIP0>;
+			ipmmu-id = <RCAR_IPMMU_VIP0>;
+		};
+
+		irq_stv_00_01 {
+			compatible = "generic-uio";
+			reg = <0 0xe7a00000 0 0x10000>;
+			interrupts = <GIC_SPI 239 IRQ_TYPE_LEVEL_HIGH>;
+			clocks = <&cpg CPG_MOD 101>;
+			power-domains = <&sysc R8A77980_PD_A3VIP0>;
+		};
+
+		dof_00 {
+			compatible = "generic-uio";
+			reg = <0 0xe7a10000 0 0x10000>;
+			interrupts = <GIC_SPI 242 IRQ_TYPE_LEVEL_HIGH>;
+			clocks = <&cpg CPG_MOD 102>;
+			power-domains = <&sysc R8A77980_PD_A3VIP1>;
+			ipmmu-id = <RCAR_IPMMU_VIP0>;
+		};
+
+		irq_dof_00_01 {
+			compatible = "generic-uio";
+			reg = <0 0xe7a10000 0 0x10000>;
+			interrupts = <GIC_SPI 243 IRQ_TYPE_LEVEL_HIGH>;
+			clocks = <&cpg CPG_MOD 102>;
+			power-domains = <&sysc R8A77980_PD_A3VIP1>;
+		};
+
+		acf_00 {
+			compatible = "generic-uio";
+			reg = <0 0xe7a50000 0 0x10000>;
+			interrupts = <GIC_SPI 71 IRQ_TYPE_LEVEL_HIGH>;
+			clocks = <&cpg CPG_MOD 1004>;
+			power-domains = <&sysc R8A77980_PD_A3VIP2>;
+			ipmmu-id = <RCAR_IPMMU_VIP1>;
+		};
+
+		irq_acf_00_01 {
+			compatible = "generic-uio";
+			reg = <0 0xe7a50000 0 0x10000>;
+			interrupts = <GIC_SPI 97 IRQ_TYPE_LEVEL_HIGH>;
+			clocks = <&cpg CPG_MOD 1004>;
+			power-domains = <&sysc R8A77980_PD_A3VIP2>;
+		};
+
+		acf_01 {
+			compatible = "generic-uio";
+			reg = <0 0xe7a60000 0 0x10000>;
+			interrupts = <GIC_SPI 98 IRQ_TYPE_LEVEL_HIGH>;
+			clocks = <&cpg CPG_MOD 1003>;
+			power-domains = <&sysc R8A77980_PD_A3VIP2>;
+			ipmmu-id = <RCAR_IPMMU_VIP1>;
+		};
+
+		irq_acf_01_01 {
+			compatible = "generic-uio";
+			reg = <0 0xe7a60000 0 0x10000>;
+			interrupts = <GIC_SPI 99 IRQ_TYPE_LEVEL_HIGH>;
+			clocks = <&cpg CPG_MOD 1003>;
+			power-domains = <&sysc R8A77980_PD_A3VIP2>;
+		};
+
+		acf_02 {
+			compatible = "generic-uio";
+			reg = <0 0xe7a70000 0 0x10000>;
+			interrupts = <GIC_SPI 100 IRQ_TYPE_LEVEL_HIGH>;
+			clocks = <&cpg CPG_MOD 1002>;
+			power-domains = <&sysc R8A77980_PD_A3VIP2>;
+			ipmmu-id = <RCAR_IPMMU_VIP1>;
+		};
+
+		irq_acf_02_01 {
+			compatible = "generic-uio";
+			reg = <0 0xe7a70000 0 0x10000>;
+			interrupts = <GIC_SPI 101 IRQ_TYPE_LEVEL_HIGH>;
+			clocks = <&cpg CPG_MOD 1002>;
+			power-domains = <&sysc R8A77980_PD_A3VIP2>;
+		};
+
+		acf_03 {
+			compatible = "generic-uio";
+			reg = <0 0xe7a80000 0 0x10000>;
+			interrupts = <GIC_SPI 102 IRQ_TYPE_LEVEL_HIGH>;
+			clocks = <&cpg CPG_MOD 1001>;
+			power-domains = <&sysc R8A77980_PD_A3VIP2>;
+			ipmmu-id = <RCAR_IPMMU_VIP1>;
+		};
+
+		irq_acf_03_01 {
+			compatible = "generic-uio";
+			reg = <0 0xe7a80000 0 0x10000>;
+			interrupts = <GIC_SPI 103 IRQ_TYPE_LEVEL_HIGH>;
+			clocks = <&cpg CPG_MOD 1001>;
+			power-domains = <&sysc R8A77980_PD_A3VIP2>;
+		};
+
+		acf_04 {
+			compatible = "generic-uio";
+			reg = <0 0xe7a90000 0 0x10000>;
+			interrupts = <GIC_SPI 104 IRQ_TYPE_LEVEL_HIGH>;
+			clocks = <&cpg CPG_MOD 1000>;
+			power-domains = <&sysc R8A77980_PD_A3VIP2>;
+			ipmmu-id = <RCAR_IPMMU_VIP1>;
+		};
+
+		irq_acf_04_01 {
+			compatible = "generic-uio";
+			reg = <0 0xe7a90000 0 0x10000>;
+			interrupts = <GIC_SPI 108 IRQ_TYPE_LEVEL_HIGH>;
+			clocks = <&cpg CPG_MOD 1000>;
+			power-domains = <&sysc R8A77980_PD_A3VIP2>;
+		};
+
+		fcpr_vip_00 {
+			compatible = "generic-uio";
+			reg = <0 0xe7aa0000 0 0x20000>;
+		};
+
+		fcpr_ir_00 {
+			compatible = "generic-uio";
+			reg = <0 0xff8d8000 0 0x10000>;
+		};
+
+		fcpr_vio_00 {
+			compatible = "generic-uio";
+			reg = <0 0xfe9c0000 0 0x10000>;
+		};
+
+		fcpr_vio_01 {
+			compatible = "generic-uio";
+			reg = <0 0xfea90000 0 0x10000>;
+		};
+
+		fcpr_rt_00 {
+			compatible = "generic-uio";
+			reg = <0 0xffc70000 0 0x10000>;
+		};
+
+		fcpr_peri_00 {
+			compatible = "generic-uio";
+			reg = <0 0xe7320000 0 0x10000>;
+		};
+
+		fcpra_peri_00 {
+			compatible = "generic-uio";
+			reg = <0 0xe7410000 0 0x10000>;
+		};
+
+		fcpr_c53_00 {
+			compatible = "generic-uio";
+			reg = <0 0xf12a0000 0 0x10000>;
+		};
+
+		fcpra_c53_00 {
+			compatible = "generic-uio";
+			reg = <0 0xf12f0000 0 0x10000>;
+		};
+
+
+		isp0: isp@fec00000 {
+			compatible = "renesas,isp-r8a77980";
+			reg = <0 0xfec00000 0 0x20000>,
+			      <0 0xfed00000 0 0x10000>;
+			interrupts = <GIC_SPI 26 IRQ_TYPE_LEVEL_HIGH>,
+				     <GIC_SPI 25 IRQ_TYPE_LEVEL_HIGH>;
+			clocks = <&cpg CPG_MOD 817>;
+			power-domains = <&sysc R8A77980_PD_ALWAYS_ON>;
+		};
+
+		isp1: isp@fee00000 {
+			compatible = "renesas,isp-r8a77980";
+			reg = <0 0xfee00000 0 0x20000>,
+			      <0 0xfed20000 0 0x10000>;
+			interrupts = <GIC_SPI 109 IRQ_TYPE_LEVEL_HIGH>,
+				     <GIC_SPI 27 IRQ_TYPE_LEVEL_HIGH>;
+			clocks = <&cpg CPG_MOD 814>;
+			power-domains = <&sysc R8A77980_PD_ALWAYS_ON>;
+		};
+
+		tisp0: tisp_00 {
+			compatible = "generic-uio";
+			reg = <0 0xfed00000 0 0x10000>;
+			interrupts = <GIC_SPI 25 IRQ_TYPE_LEVEL_HIGH>;
+			clocks = <&cpg CPG_MOD 817>;
+			resets = <&cpg 817>;
+			power-domains = <&sysc R8A77980_PD_ALWAYS_ON>;
+			ipmmu-id = <RCAR_IPMMU_VI0>;
+		};
+
+		cisp0: cisp_00 {
+			compatible = "generic-uio";
+			reg = <0 0xfec00000 0 0x20000>;
+			interrupts = <GIC_SPI 26 IRQ_TYPE_LEVEL_HIGH>;
+			clocks = <&cpg CPG_MOD 817>;
+			resets = <&cpg 817>;
+			power-domains = <&sysc R8A77980_PD_ALWAYS_ON>;
+			ipmmu-id = <RCAR_IPMMU_VI0>;
+		};
+
+		tisp1: tisp_01 {
+			compatible = "generic-uio";
+			reg = <0 0xfed20000 0 0x10000>;
+			interrupts = <GIC_SPI 27 IRQ_TYPE_LEVEL_HIGH>;
+			clocks = <&cpg CPG_MOD 814>;
+			resets = <&cpg 814>;
+			power-domains = <&sysc R8A77980_PD_ALWAYS_ON>;
+			ipmmu-id = <RCAR_IPMMU_VI0>;
+		};
+
+		cisp1: cisp_01 {
+			compatible = "generic-uio";
+			reg = <0 0xfee00000 0 0x20000>;
+			interrupts = <GIC_SPI 109 IRQ_TYPE_LEVEL_HIGH>;
+			clocks = <&cpg CPG_MOD 814>;
+			resets = <&cpg 814>;
+			power-domains = <&sysc R8A77980_PD_ALWAYS_ON>;
+			ipmmu-id = <RCAR_IPMMU_VI0>;
 		};
 
 		dmac1: dma-controller@e7300000 {
@@ -1327,9 +1828,9 @@
 			#iommu-cells = <1>;
 		};
 
-		ipmmu_vc0: mmu@fe990000 {
+		ipmmu_vc0: mmu@fe6b0000 {
 			compatible = "renesas,ipmmu-r8a77980";
-			reg = <0 0xfe990000 0 0x1000>;
+			reg = <0 0xfe6b0000 0 0x1000>;
 			renesas,ipmmu-main = <&ipmmu_mm 12>;
 			power-domains = <&sysc R8A77980_PD_ALWAYS_ON>;
 			#iommu-cells = <1>;
@@ -1420,8 +1921,8 @@
 			status = "disabled";
 		};
 
-		vspd0: vsp@fea20000 {
-			compatible = "renesas,vsp2";
+		vspd0: vsp_00@fea20000 {
+			compatible = "renesas,vsp2", "generic-uio";
 			reg = <0 0xfea20000 0 0x5000>;
 			interrupts = <GIC_SPI 169 IRQ_TYPE_LEVEL_HIGH>;
 			clocks = <&cpg CPG_MOD 623>;
@@ -1438,14 +1939,14 @@
 			resets = <&cpg 603>;
 		};
 
-		csi40: csi2@feaa0000 {
-			compatible = "renesas,r8a77980-csi2";
+		csi40: csi_00@feaa0000 {
+			compatible = "renesas,r8a77980-csi2", "generic-uio";
 			reg = <0 0xfeaa0000 0 0x10000>;
 			interrupts = <GIC_SPI 246 IRQ_TYPE_LEVEL_HIGH>;
 			clocks = <&cpg CPG_MOD 716>;
 			power-domains = <&sysc R8A77980_PD_ALWAYS_ON>;
 			resets = <&cpg 716>;
-			status = "disabled";
+			status = "enabled";
 
 			ports {
 				#address-cells = <1>;
@@ -1477,14 +1978,14 @@
 			};
 		};
 
-		csi41: csi2@feab0000 {
-			compatible = "renesas,r8a77980-csi2";
+		csi41: csi_01@feab0000 {
+			compatible = "renesas,r8a77980-csi2", "generic-uio";
 			reg = <0 0xfeab0000 0 0x10000>;
 			interrupts = <GIC_SPI 241 IRQ_TYPE_LEVEL_HIGH>;
 			clocks = <&cpg CPG_MOD 715>;
 			power-domains = <&sysc R8A77980_PD_ALWAYS_ON>;
 			resets = <&cpg 715>;
-			status = "disabled";
+			status = "enabled";
 
 			ports {
 				#address-cells = <1>;
@@ -1516,8 +2017,8 @@
 			};
 		};
 
-		du: display@feb00000 {
-			compatible = "renesas,du-r8a77980",
+		du: du_00@feb00000 {
+			compatible = "renesas,du-r8a77980", "generic-uio",
 				     "renesas,du-r8a77970";
 			reg = <0 0xfeb00000 0 0x80000>;
 			interrupts = <GIC_SPI 256 IRQ_TYPE_LEVEL_HIGH>;
@@ -1547,9 +2048,9 @@
 			};
 		};
 
-		lvds0: lvds-encoder@feb90000 {
-			compatible = "renesas,r8a77980-lvds";
-			reg = <0 0xfeb90000 0 0x14>;
+		lvds0: lvds_00@feb90000 {
+			compatible = "renesas,r8a77980-lvds", "generic-uio";
+			reg = <0 0xfeb90000 0 0x1000>;
 			clocks = <&cpg CPG_MOD 727>;
 			power-domains = <&sysc R8A77980_PD_ALWAYS_ON>;
 			resets = <&cpg 727>;
@@ -1579,6 +2080,409 @@
 			compatible = "renesas,prr";
 			reg = <0 0xfff00044 0 4>;
 		};
+
+		iccom_00 {
+			compatible = "renesas,iccom-rcar", "generic-uio";
+			interrupts = <0 224 4>;
+			reg = <0x0 0xe6260000 0x0 0x1000>;
+			iccom,reg-offset = <0x400 0x404 0x440 0x460>;
+			iccom,cta-memory = <0x0 0x47fc7000 0x0 0x2000>;
+			iccom,ack-timeout = <100>;
+			iccom,trg-timeout = <100>;
+		};
+
+		irq_iccom_00_01 {
+			compatible = "renesas,iccom-rcar", "generic-uio";
+			interrupts = <0 225 4>;
+			reg = <0x0 0xe6260000 0x0 0x1000>;
+			iccom,reg-offset = <0x408 0x40c 0x444 0x464>;
+			iccom,cta-memory = <0x0 0x47fc9000 0x0 0x2000>;
+			iccom,ack-timeout = <100>;
+			iccom,trg-timeout = <100>;
+		};
+
+		irq_iccom_00_02 {
+			compatible = "renesas,iccom-rcar", "generic-uio";
+			interrupts = <0 226 4>;
+			reg = <0x0 0xe6260000 0x0 0x1000>;
+			iccom,reg-offset = <0x410 0x414 0x448 0x468>;
+			iccom,cta-memory = <0x0 0x47fcb000 0x0 0x2000>;
+			iccom,ack-timeout = <100>;
+			iccom,trg-timeout = <100>;
+		};
+
+		irq_iccom_00_03 {
+			compatible = "renesas,iccom-rcar", "generic-uio";
+			interrupts = <0 227 4>;
+			reg = <0x0 0xe6260000 0x0 0x1000>;
+			iccom,reg-offset = <0x418 0x41c 0x44c 0x46c>;
+			iccom,cta-memory = <0x0 0x47fcd000 0x0 0x2000>;
+			iccom,ack-timeout = <100>;
+			iccom,trg-timeout = <100>;
+		};
+
+		irq_iccom_00_04 {
+			compatible = "renesas,iccom-rcar", "generic-uio";
+			interrupts = <0 228 4>;
+			reg = <0x0 0xe6260000 0x0 0x1000>;
+			iccom,reg-offset = <0x420 0x424 0x450 0x470>;
+			iccom,cta-memory = <0x0 0x47fcf000 0x0 0x2000>;
+			iccom,ack-timeout = <100>;
+			iccom,trg-timeout = <100>;
+		};
+
+		irq_iccom_00_05 {
+			compatible = "renesas,iccom-rcar", "generic-uio";
+			interrupts = <0 229 4>;
+			reg = <0x0 0xe6260000 0x0 0x1000>;
+			iccom,reg-offset = <0x428 0x42c 0x454 0x474>;
+			iccom,cta-memory = <0x0 0x47fd1000 0x0 0x2000>;
+			iccom,ack-timeout = <100>;
+			iccom,trg-timeout = <100>;
+		};
+
+		irq_iccom_00_06 {
+			compatible = "renesas,iccom-rcar", "generic-uio";
+			interrupts = <0 230 4>;
+			reg = <0x0 0xe6260000 0x0 0x1000>;
+			iccom,reg-offset = <0x430 0x434 0x458 0x478>;
+			iccom,cta-memory = <0x0 0x47fd3000 0x0 0x2000>;
+			iccom,ack-timeout = <100>;
+			iccom,trg-timeout = <100>;
+		};
+
+		irq_iccom_00_07 {
+			compatible = "renesas,iccom-rcar", "generic-uio";
+			interrupts = <0 231 4>;
+			reg = <0x0 0xe6260000 0x0 0x1000>;
+			iccom,reg-offset = <0x438 0x43c 0x45c 0x47c>;
+			iccom,cta-memory = <0x0 0x47fd5000 0x0 0x2000>;
+			iccom,ack-timeout = <100>;
+			iccom,trg-timeout = <100>;
+		};
+
+		irq_iccom_00_08 { /* iccom error 0 */
+			compatible = "generic-uio";
+			interrupts = <GIC_SPI 337 IRQ_TYPE_LEVEL_HIGH>;
+			reg = <0x0 0xe6260000 0x0 0x1000>;
+			power-domains = <&sysc R8A77980_PD_ALWAYS_ON>;
+		};
+
+		irq_iccom_00_09 { /* iccom error 1 */
+			compatible = "generic-uio";
+			interrupts = <GIC_SPI 338 IRQ_TYPE_LEVEL_HIGH>;
+			reg = <0x0 0xe6260000 0x0 0x1000>;
+			power-domains = <&sysc R8A77980_PD_ALWAYS_ON>;
+		};
+
+		irq_iccom_00_10 { /* iccom error 2 */
+			compatible = "generic-uio";
+			interrupts = <GIC_SPI 339 IRQ_TYPE_LEVEL_HIGH>;
+			reg = <0x0 0xe6260000 0x0 0x1000>;
+			power-domains = <&sysc R8A77980_PD_ALWAYS_ON>;
+		};
+
+		irq_iccom_00_11 { /* iccom error 3 */
+			compatible = "generic-uio";
+			interrupts = <GIC_SPI 340 IRQ_TYPE_LEVEL_HIGH>;
+			reg = <0x0 0xe6260000 0x0 0x1000>;
+			power-domains = <&sysc R8A77980_PD_ALWAYS_ON>;
+		};
+
+		irq_iccom_00_12 { /* iccom error 4 */
+			compatible = "generic-uio";
+			interrupts = <GIC_SPI 341 IRQ_TYPE_LEVEL_HIGH>;
+			reg = <0x0 0xe6260000 0x0 0x1000>;
+			power-domains = <&sysc R8A77980_PD_ALWAYS_ON>;
+		};
+
+		irq_iccom_00_13 { /* iccom error 5 */
+			compatible = "generic-uio";
+			interrupts = <GIC_SPI 342 IRQ_TYPE_LEVEL_HIGH>;
+			reg = <0x0 0xe6260000 0x0 0x1000>;
+			power-domains = <&sysc R8A77980_PD_ALWAYS_ON>;
+		};
+
+		irq_iccom_00_14 { /* iccom error 6 */
+			compatible = "generic-uio";
+			interrupts = <GIC_SPI 343 IRQ_TYPE_LEVEL_HIGH>;
+			reg = <0x0 0xe6260000 0x0 0x1000>;
+			power-domains = <&sysc R8A77980_PD_ALWAYS_ON>;
+		};
+
+		irq_iccom_00_15 { /* iccom error 7 */
+			compatible = "generic-uio";
+			interrupts = <GIC_SPI 344 IRQ_TYPE_LEVEL_HIGH>;
+			reg = <0x0 0xe6260000 0x0 0x1000>;
+			power-domains = <&sysc R8A77980_PD_ALWAYS_ON>;
+		};
+
+		irq_iccom_00_16 { /* iccom error 8 */
+			compatible = "generic-uio";
+			interrupts = <GIC_SPI 31 IRQ_TYPE_LEVEL_HIGH>;
+			reg = <0x0 0xe6260000 0x0 0x1000>;
+			power-domains = <&sysc R8A77980_PD_ALWAYS_ON>;
+		};
+
+		irq_iccom_00_17 { /* iccom error 9 */
+			compatible = "generic-uio";
+			interrupts = <GIC_SPI 347 IRQ_TYPE_LEVEL_HIGH>;
+			reg = <0x0 0xe6260000 0x0 0x1000>;
+			power-domains = <&sysc R8A77980_PD_ALWAYS_ON>;
+		};
+
+		irq_iccom_00_18 { /* iccom error 10 */
+			compatible = "generic-uio";
+			interrupts = <GIC_SPI 348 IRQ_TYPE_LEVEL_HIGH>;
+			reg = <0x0 0xe6260000 0x0 0x1000>;
+			power-domains = <&sysc R8A77980_PD_ALWAYS_ON>;
+		};
+
+		irq_iccom_00_19 { /* iccom error 11 */
+			compatible = "generic-uio";
+			interrupts = <GIC_SPI 349 IRQ_TYPE_LEVEL_HIGH>;
+			reg = <0x0 0xe6260000 0x0 0x1000>;
+			power-domains = <&sysc R8A77980_PD_ALWAYS_ON>;
+		};
+
+		irq_iccom_00_20 { /* iccom error 12 */
+			compatible = "generic-uio";
+			interrupts = <GIC_SPI 352 IRQ_TYPE_LEVEL_HIGH>;
+			reg = <0x0 0xe6260000 0x0 0x1000>;
+			power-domains = <&sysc R8A77980_PD_ALWAYS_ON>;
+		};
+
+		irq_iccom_00_21 { /* iccom error 13 */
+			compatible = "generic-uio";
+			interrupts = <GIC_SPI 376 IRQ_TYPE_LEVEL_HIGH>;
+			reg = <0x0 0xe6260000 0x0 0x1000>;
+			power-domains = <&sysc R8A77980_PD_ALWAYS_ON>;
+		};
+
+		qos@e67e0000 {
+			compatible = "renesas,qos";
+			reg = <0 0xe67e0000 0 0x10090>;
+		};
+
+		ivcp1e_00 {
+			compatible = "generic-uio";
+			reg = <0 0xfea00000 0 0x10000>;
+			clocks = <&cpg CPG_MOD 127>;
+			power-domains = <&sysc R8A77980_PD_ALWAYS_ON>;
+			interrupts = <GIC_SPI 300 IRQ_TYPE_LEVEL_HIGH>;
+			ipmmu-id = <RCAR_IPMMU_VI0>;
+		};
+
+		irq_ivcp1e_00_01 {
+			compatible = "generic-uio";
+			reg = <0 0xfea00000 0 0x10000>;
+			clocks = <&cpg CPG_MOD 127>;
+			power-domains = <&sysc R8A77980_PD_ALWAYS_ON>;
+			interrupts = <GIC_SPI 170 IRQ_TYPE_LEVEL_HIGH>;
+		};
+
+		rtt_fbc_post {
+			compatible = "generic-uio";
+			reg = <0 0xff830000 0 0x10000>;
+			clocks = <&cpg CPG_CORE R8A77980_CLK_POST>;
+			power-domains = <&sysc R8A77980_PD_ALWAYS_ON>;
+		};
+
+		rtt_fbc_post2 {
+			compatible = "generic-uio";
+			reg = <0 0xff830000 0 0x10000>;
+			clocks = <&cpg CPG_CORE R8A77980_CLK_POST2>;
+			power-domains = <&sysc R8A77980_PD_ALWAYS_ON>;
+		};
+
+		rtt_fbc_post4 {
+			compatible = "generic-uio";
+			reg = <0 0xff830000 0 0x10000>;
+			clocks = <&cpg CPG_CORE R8A77980_CLK_POST4>;
+			power-domains = <&sysc R8A77980_PD_ALWAYS_ON>;
+		};
+
+		rtt_fbc_finish {
+			compatible = "generic-uio";
+			reg = <0 0xff830000 0 0x10000>;
+			power-domains = <&sysc R8A77980_PD_ALWAYS_ON>;
+			interrupts = <GIC_SPI 14 IRQ_TYPE_LEVEL_HIGH>;
+		};
+
+		rtt_fbc_pre {
+			compatible = "generic-uio";
+			reg = <0 0xff830000 0 0x10000>;
+			power-domains = <&sysc R8A77980_PD_ALWAYS_ON>;
+			interrupts = <GIC_SPI 15 IRQ_TYPE_LEVEL_HIGH>;
+		};
+
+		imr_rtt_imr_00 {
+			compatible = "generic-uio";
+			reg = <0 0xfff80000 0 0x1000>;
+			clocks = <&cpg CPG_MOD 707>;
+			power-domains = <&sysc R8A77980_PD_ALWAYS_ON>;
+		};
+
+		imr_rtt_imr_01 {
+			compatible = "generic-uio";
+			reg = <0 0xfff81000 0 0x1000>;
+			clocks = <&cpg CPG_MOD 706>;
+			power-domains = <&sysc R8A77980_PD_ALWAYS_ON>;
+		};
+
+		imr_rtt_ims_00 {
+			compatible = "generic-uio";
+			reg = <0 0xfff90000 0 0x1000>;
+			clocks = <&cpg CPG_MOD 823>;
+			power-domains = <&sysc R8A77980_PD_ALWAYS_ON>;
+		};
+
+		imr_rtt_ims_01 {
+			compatible = "generic-uio";
+			reg = <0 0xfff91000 0 0x1000>;
+			clocks = <&cpg CPG_MOD 822>;
+			power-domains = <&sysc R8A77980_PD_ALWAYS_ON>;
+		};
+
+		imr_rtt_ims_02 {
+			compatible = "generic-uio";
+			reg = <0 0xfff92000 0 0x1000>;
+			clocks = <&cpg CPG_MOD 821>;
+			power-domains = <&sysc R8A77980_PD_ALWAYS_ON>;
+		};
+
+		imr_rtt_ims_03 {
+			compatible = "generic-uio";
+			reg = <0 0xfff93000 0 0x1000>;
+			clocks = <&cpg CPG_MOD 820>;
+			power-domains = <&sysc R8A77980_PD_ALWAYS_ON>;
+		};
+
+		viprtt_dof_00 {
+			compatible = "generic-uio";
+			reg = <0 0xe7930000 0 0x1000>;
+			clocks = <&cpg CPG_MOD 102>;
+			power-domains = <&sysc R8A77980_PD_A3VIP1>;
+		};
+
+		viprtt_stv_00 {
+			compatible = "generic-uio";
+			reg = <0 0xe7950000 0 0x1000>;
+			clocks = <&cpg CPG_MOD 101>;
+			power-domains = <&sysc R8A77980_PD_A3VIP0>;
+		};
+
+		viprtt_acf_00 {
+			compatible = "generic-uio";
+			reg = <0 0xe7940000 0 0x1000>;
+			clocks = <&cpg CPG_MOD 1004>;
+			power-domains = <&sysc R8A77980_PD_A3VIP2>;
+		};
+
+		viprtt_acf_01 {
+			compatible = "generic-uio";
+			reg = <0 0xe7941000 0 0x1000>;
+			clocks = <&cpg CPG_MOD 1002>;
+			power-domains = <&sysc R8A77980_PD_A3VIP2>;
+		};
+
+		imp_rtt_imp_00 {
+			compatible = "generic-uio";
+			reg = <0 0xff8d0000 0 0x1000>;
+			clocks = <&cpg CPG_MOD 827>;
+			power-domains = <&sysc R8A77980_PD_A2IR0>;
+		};
+
+		imp_rtt_imp_01 {
+			compatible = "generic-uio";
+			reg = <0 0xfff50000 0 0x1000>;
+			clocks = <&cpg CPG_MOD 826>;
+			power-domains = <&sysc R8A77980_PD_A2IR1>;
+		};
+
+		imp_rtt_imp_02 {
+			compatible = "generic-uio";
+			reg = <0 0xfff60000 0 0x1000>;
+			clocks = <&cpg CPG_MOD 825>;
+			power-domains = <&sysc R8A77980_PD_A2IR2>;
+		};
+
+		imp_rtt_imp_03 {
+			compatible = "generic-uio";
+			reg = <0 0xfff70000 0 0x1000>;
+			clocks = <&cpg CPG_MOD 824>;
+			power-domains = <&sysc R8A77980_PD_A2IR3>;
+		};
+
+		imp_rtt_imp_04 {
+			compatible = "generic-uio";
+			reg = <0 0xff8d1000 0 0x1000>;
+			clocks = <&cpg CPG_MOD 521>;
+			power-domains = <&sysc R8A77980_PD_A2IR4>;
+		};
+
+		imp_rtt_ocv_00 {
+			compatible = "generic-uio";
+			reg = <0 0xfffa0000 0 0x1000>;
+			clocks = <&cpg CPG_MOD 829>;
+			power-domains = <&sysc R8A77980_PD_A2SC0>;
+		};
+
+		imp_rtt_ocv_01 {
+			compatible = "generic-uio";
+			reg = <0 0xfffb0000 0 0x1000>;
+			clocks = <&cpg CPG_MOD 828>;
+			power-domains = <&sysc R8A77980_PD_A2SC1>;
+		};
+
+		imp_rtt_ocv_02 {
+			compatible = "generic-uio";
+			reg = <0 0xff8d2000 0 0x1000>;
+			clocks = <&cpg CPG_MOD 531>;
+			power-domains = <&sysc R8A77980_PD_A2SC2>;
+		};
+
+		imp_rtt_ocv_03 {
+			compatible = "generic-uio";
+			reg = <0 0xff8d3000 0 0x1000>;
+			clocks = <&cpg CPG_MOD 529>;
+			power-domains = <&sysc R8A77980_PD_A2SC3>;
+		};
+
+		imp_rtt_ocv_04 {
+			compatible = "generic-uio";
+			reg = <0 0xff8d4000 0 0x1000>;
+			clocks = <&cpg CPG_MOD 528>;
+			power-domains = <&sysc R8A77980_PD_A2SC4>;
+		};
+
+		imp_rtt_dp_00 {
+			compatible = "generic-uio";
+			reg = <0 0xff8d5000 0 0x1000>;
+			clocks = <&cpg CPG_MOD 525>;
+			power-domains = <&sysc R8A77980_PD_A2DP0>;
+		};
+
+		imp_rtt_dp_01 {
+			compatible = "generic-uio";
+			reg = <0 0xff8d6000 0 0x1000>;
+			clocks = <&cpg CPG_MOD 524>;
+			power-domains = <&sysc R8A77980_PD_A2DP1>;
+		};
+
+		imp_rtt_cnn_00 {
+			compatible = "generic-uio";
+			reg = <0 0xff8d7000 0 0x1000>;
+			clocks = <&cpg CPG_MOD 831>;
+			power-domains = <&sysc R8A77980_PD_A2CN>;
+		};
+
+		imp_rtt_slim_00 {
+			compatible = "generic-uio";
+			reg = <0 0xff8dc000 0 0x1000>;
+			clocks = <&cpg CPG_MOD 500>;
+			power-domains = <&sysc R8A77980_PD_A2IR5>;
+		};
 	};
 
 	thermal-zones {
diff --git a/drivers/clk/renesas/r8a77970-cpg-mssr.c b/drivers/clk/renesas/r8a77970-cpg-mssr.c
index 05917f6..ccbafee 100644
--- a/drivers/clk/renesas/r8a77970-cpg-mssr.c
+++ b/drivers/clk/renesas/r8a77970-cpg-mssr.c
@@ -29,7 +29,7 @@ enum r8a77970_clk_types {
 
 enum clk_ids {
 	/* Core Clock Outputs exported to DT */
-	LAST_DT_CORE_CLK = R8A77970_CLK_OSC,
+	LAST_DT_CORE_CLK = R8A77970_CLK_POST2,
 
 	/* External Input Clocks */
 	CLK_EXTAL,
@@ -102,6 +102,8 @@ static const struct cpg_core_clk r8a77970_core_clks[] __initconst = {
 	DEF_DIV6P1("canfd",	R8A77970_CLK_CANFD, CLK_PLL1_DIV4, 0x244),
 	DEF_DIV6P1("mso",	R8A77970_CLK_MSO,   CLK_PLL1_DIV4, 0x014),
 	DEF_DIV6P1("csi0",	R8A77970_CLK_CSI0,  CLK_PLL1_DIV4, 0x00c),
+	DEF_DIV6P1("post",	R8A77970_CLK_POST,  CLK_PLL1_DIV4, 0x08c),
+	DEF_DIV6P1("post2",	R8A77970_CLK_POST2, CLK_PLL1_DIV4, 0x09c),
 
 	DEF_FIXED("osc",	R8A77970_CLK_OSC,   CLK_PLL1_DIV2, 12*1024, 1),
 	DEF_FIXED("r",		R8A77970_CLK_R,	    CLK_EXTALR,	   1, 1),
diff --git a/drivers/clk/renesas/r8a77980-cpg-mssr.c b/drivers/clk/renesas/r8a77980-cpg-mssr.c
index 50789c9..01ea098 100644
--- a/drivers/clk/renesas/r8a77980-cpg-mssr.c
+++ b/drivers/clk/renesas/r8a77980-cpg-mssr.c
@@ -23,7 +23,7 @@
 
 enum clk_ids {
 	/* Core Clock Outputs exported to DT */
-	LAST_DT_CORE_CLK = R8A77980_CLK_OSC,
+	LAST_DT_CORE_CLK = R8A77980_CLK_POST4,
 
 	/* External Input Clocks */
 	CLK_EXTAL,
@@ -106,6 +106,9 @@ static const struct cpg_core_clk r8a77980_core_clks[] __initconst = {
 	DEF_DIV6P1("canfd",	R8A77980_CLK_CANFD, CLK_PLL1_DIV4, 0x244),
 	DEF_DIV6P1("csi0",	R8A77980_CLK_CSI0,  CLK_PLL1_DIV4, 0x00c),
 	DEF_DIV6P1("mso",	R8A77980_CLK_MSO,   CLK_PLL1_DIV4, 0x014),
+	DEF_DIV6P1("post",	R8A77980_CLK_POST,  CLK_PLL1_DIV4, 0x08c),
+	DEF_DIV6P1("post2",	R8A77980_CLK_POST2, CLK_PLL1_DIV4, 0x09c),
+	DEF_DIV6P1("post4",	R8A77980_CLK_POST4, CLK_PLL1_DIV4, 0x260),
 
 	DEF_GEN3_OSC("osc",	R8A77980_CLK_OSC,   CLK_EXTAL,     8),
 	DEF_GEN3_MDSEL("r",	R8A77980_CLK_R, 29, CLK_EXTALR, 1, CLK_OCO, 1),
@@ -119,6 +122,7 @@ static const struct mssr_mod_clk r8a77980_mod_clks[] __initconst = {
 	DEF_MOD("tmu2",			 123,	R8A77980_CLK_S0D6),
 	DEF_MOD("tmu1",			 124,	R8A77980_CLK_S0D6),
 	DEF_MOD("tmu0",			 125,	R8A77980_CLK_CP),
+	DEF_MOD("ivcp1e",		 127,	R8A77980_CLK_S2D1),
 	DEF_MOD("scif4",		 203,	R8A77980_CLK_S3D4),
 	DEF_MOD("scif3",		 204,	R8A77980_CLK_S3D4),
 	DEF_MOD("scif1",		 206,	R8A77980_CLK_S3D4),
diff --git a/include/dt-bindings/clock/r8a77970-cpg-mssr.h b/include/dt-bindings/clock/r8a77970-cpg-mssr.h
index 6145ebe..2d2b01b 100644
--- a/include/dt-bindings/clock/r8a77970-cpg-mssr.h
+++ b/include/dt-bindings/clock/r8a77970-cpg-mssr.h
@@ -40,5 +40,7 @@
 #define R8A77970_CLK_CPEX		28
 #define R8A77970_CLK_R			29
 #define R8A77970_CLK_OSC		30
+#define R8A77970_CLK_POST		31
+#define R8A77970_CLK_POST2		32
 
 #endif /* __DT_BINDINGS_CLOCK_R8A77970_CPG_MSSR_H__ */
diff --git a/include/dt-bindings/clock/r8a77980-cpg-mssr.h b/include/dt-bindings/clock/r8a77980-cpg-mssr.h
index a4c0d76..bb17430 100644
--- a/include/dt-bindings/clock/r8a77980-cpg-mssr.h
+++ b/include/dt-bindings/clock/r8a77980-cpg-mssr.h
@@ -47,5 +47,8 @@
 #define R8A77980_CLK_CPEX		35
 #define R8A77980_CLK_R			36
 #define R8A77980_CLK_OSC		37
+#define R8A77980_CLK_POST		38
+#define R8A77980_CLK_POST2		39
+#define R8A77980_CLK_POST4		40
 
 #endif /* __DT_BINDINGS_CLOCK_R8A77980_CPG_MSSR_H__ */
-- 
2.7.4

